p2p/pubsub/psHost.go

```diff
@@ -0,0 +1,366 @@
+		return err
+	}
+
+	psh.handlers[specifier] = handler
+	return nil
+}
+
+// startHandler start the handler with the given specifier.
+// If the topic of the handler is not running, run it.
+func (psh *pubSubHost) startHandler(spec HandlerSpecifier) error {
+	handler, exist := psh.handlers[spec]
+	if !exist {
+		return errHandlerNotExist
+	}
+	tr, exist := psh.topicRunners[handler.Topic()]
+	if !exist {
+		return errTopicNotRegistered
+	}
+
+	if err := tr.addHandler(handler); err != nil {
+		return err
+	}
+
+	if !tr.isRunning() {
+		psh.log.Info().Str("topic", string(tr.topic)).Msg("starting topic")
+		if err := tr.start(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// stopHandler stop the handler with the given specifier.
+// If the topic of the handler has no more handlers to play with, stop the handler.
+func (psh *pubSubHost) stopHandler(spec HandlerSpecifier) error {
+	handler, exist := psh.handlers[spec]
+	if !exist {
+		return errHandlerNotExist
+	}
+	tr, err := psh.getTopicRunner(handler.Topic())
+	if err != nil {
+		return err
+	}
+	if err := tr.removeHandler(spec); err != nil {
+		return err
+	}
+	if tr.hasNoHandlerRunning() {
+		psh.log.Info().Str("topic", string(handler.Topic())).Msg("stopping topic")
+		if err := tr.stop(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// removeHandler remove the handler with the given specifier.
+// If the topic of the handler has no more handlers (running and not running),
+// stop listen the topic.
+func (psh *pubSubHost) removeHandler(spec HandlerSpecifier) error {
+	handler, exist := psh.handlers[spec]
+	if !exist {
+		return errHandlerNotExist
+	}
+	topic := handler.Topic()
+	tr, err := psh.getTopicRunner(topic)
+	if err != nil {
+		return err
+	}
+
+	if err := tr.removeHandler(spec); err != nil {
+		if !errors.Is(err, errHandlerNotExist) {
+			return err
+		}
+	}
+
+	delete(psh.handlers, spec)
+	if psh.haveNoHandlerOnTopic(topic) {
+		return psh.closeAndRemoveTopicRunner(topic)
+	}
+	if tr.hasNoHandlerRunning() {
+		psh.log.Info().Str("topic", string(handler.Topic())).Msg("stopping topic")
+		if err := tr.stop(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// removeTopic close the topic and remove all handlers in the topic
+func (psh *pubSubHost) removeTopic(topic Topic) error {
+	tr, err := psh.getTopicRunner(topic)
+	if err != nil {
+		return err
+	}
+
+	psh.log.Info().Str("topic", string(topic)).Msg("closing topic")
+	if err := tr.close(); err != nil {
+		return err
+	}
+	delete(psh.topicRunners, topic)
+	psh.deleteHandlersWithTopic(topic)
+	return nil
+}
+
+func (psh *pubSubHost) haveNoHandlerOnTopic(topic Topic) bool {
+	for _, handler := range psh.handlers {
+		if handler.Topic() == topic {
+			return false
+		}
+	}
+	return true
+}
+
+func (psh *pubSubHost) deleteHandlersWithTopic(topic Topic) {
+	for spec, handler := range psh.handlers {
+		if handler.Topic() == topic {
+			delete(psh.handlers, spec)
```

Comment:
These are run in single threaded code, so no race condition right?

Reply:
Yes. All updates on the running handlers are done in a single thread. You can see the logic in `pubSubHost.run`